build: refactor pkg-config for shared libraries

jbergstroem opened the pull request:
Improve detection and usage of pkg-config. This simplifies the setup of all our shared libraries.

If pkg-config is installed on the host and `--shared` flags are passed by the user, we try to get defaults from pkg-config instead of using the default provided by configure.

/R = @iojs/build, @bnoordhuis

jbergstroem commented:
@bnoordhuis You previously suggested 'enabling' a shared library by providing any of the libpath, includedir or other flags. I still think that should be avoided. Using `pkg-config` if available should always return accurate results since that's kind of its job.

bnoordhuis commented on configure:
```diff
@@ -343,18 +343,31 @@ def b(value):
+    retval += (val,)
 
-  return (libs, cflags)
+  return retval
```
Style: two newlines after functions.

bnoordhuis commented on configure:
```diff
@@ -343,18 +343,31 @@ def b(value):
+  pkg_config = os.environ.get('PKG_CONFIG', 'pkg-config')
+  pkg_config_args = '--silence-errors'
+  try:
+    subprocess.check_output([pkg_config, '--version'])
```
Is this explicit 'is pkg-config installed?' check necessary?  Perhaps you can fold it into the subprocess call below?  The net result - a `(None, None)` tuple - should be the same.

jbergstroem replied:
My thought was to avoid one call if pkg-config wasn't installed. I'll have a go at improving this.

bnoordhuis commented on configure:
```diff
@@ -343,18 +343,31 @@ def b(value):
 
+def format_libraries(list):
+  """Returns string of space separated libraries"""
+  list = list.replace(' ', '')
```
Why are you stripping whitespace here?

jbergstroem replied:
For consistency. Guess its cosmetics.

bnoordhuis commented on configure:
```diff
@@ -672,41 +685,25 @@ def configure_node(o):
+    if libs:
+      output['libraries'] += libs.split()
+    if cflags:
+      output['include_dirs'] += cflags.split()
```
Apropos the `libs.split()` and `cflags.split()` calls, what happens when they contain paths with spaces?

bnoordhuis replied:
Also, I'm not 100% confident it's correct to assign the result of `cflags.split()` to include_dirs.

jbergstroem replied:
Good point. Split is unsafe.

bnoordhuis replied:
Looking at it again, I don't include_dirs is the right field, it should be cflags or cflags_cc.  Probably the former.

bnoordhuis commented on configure:
```diff
@@ -718,26 +715,12 @@ def configure_v8(o):
-    else:
-      o['cflags'] += cflags.split()
-
+  configure_library('openssl', o)
```
Style: two newlines.

jbergstroem replied:
How about a separate pr with a full pep8 on configure?

bnoordhuis replied:
I personally don't care that much about pep8.  As long as changes are not too incongruent with existing code, they're fine by me.

jbergstroem replied:
Ok; was just looking through the eslint PR and thought we could improve configure as well.

bnoordhuis commented on configure:
```diff
@@ -1020,15 +1003,22 @@ if (options.dest_os):
+for library in output['libraries']:
+  if library not in unique_libraries:
+     unique_libraries.append(library)
+output['libraries'] = unique_libraries
```
Maybe just `output['libraries'] = list(set(output['libraries']))`?

jbergstroem commented:
@bnoordhuis addressed all issues besides how to handle cflags. Here's a patch I'm considering seeing that `pkg-config` returns `-I/foo`, whereas we assume user just provides path:
```diff
diff --git configure configure
index def2367..8ef0b38 100755
--- configure
+++ configure
@@ -347,9 +347,9 @@ def pkg_config(pkg):
     os.environ.get('PKG_CONFIG', 'pkg-config'))
   pkg_config_args = '--silence-errors'
   retval = ()
-  for flag in ['--libs', '--cflags']:
+  for flag in ['--libs-only-l', '--cflags-only-I']:
     try:
-     val = subprocess.check_output([pkg_config, pkg_config_args, flag, pkg])
+     val = subprocess.check_output([pkg_config, pkg_config_args, flag, pkg]).rstrip('\n')
     except subprocess.CalledProcessError:
       # most likely missing a .pc-file
       val = None
@@ -691,12 +691,12 @@ def configure_library(lib, output):

     (pkg_libs, pkg_cflags) = pkg_config(lib)
     libs = pkg_libs if pkg_libs else default_lib
-    cflags = pkg_cflags if pkg_cflags else default_cflags
+    cflags = pkg_cflags.split("-I") if pkg_cflags else default_cflags

     if libs:
       output['libraries'] += libs.split()
     if cflags:
-      output['include_dirs'] += cflags.split()
+      output['include_dirs'] += [cflags]
     if default_libpath:
       output['libraries'] += ['-L%s' % default_libpath]
```

I don't particularly like that we pass flags through libraries but not in include_dirs, but that's related to gyp I guess.

jbergstroem commented:
@bnoordhuis any feedback on above? I guess I should additionally fetch `-L` from pkg-config for default_libpath?

bnoordhuis commented on configure:
```diff
@@ -343,18 +343,26 @@ def b(value):
-  return (libs, cflags)
-
+  pkg_config = os.environ.get('PKG_CONFIG_PATH',
+    os.environ.get('PKG_CONFIG', 'pkg-config'))
```
Style: arguments should line up.  Maybe it reads slightly easier as:

```python
pkg_config = 'pkg-config'
pkg_config = os.environ.get('PKG_CONFIG', pkg_config)
pkg_config = os.environ.get('PKG_CONFIG_PATH', pkg_config)
```

bnoordhuis commented on configure:
```diff
@@ -343,18 +343,26 @@ def b(value):
+    retval += (val,)
+  return retval
+
+def format_libraries(list):
```
Style: two newlines between functions.

bnoordhuis commented on configure:
```diff
@@ -672,41 +680,25 @@ def configure_node(o):
-
+def configure_library(lib, output):
+  shared_lib = 'shared_%s' % lib
+  output['variables']['node_%s' % shared_lib] = b(getattr(options, shared_lib))
```
Just a note but you can write this as `'shared_' + lib` and `'node_' + shared_lib` as well.  I'll leave it up to you to decide what's more readable.

bnoordhuis commented:
Do you happen to know when --cflags-only-I and --libs-only-L were added?  They haven't always been around but I'm not sure when they first appeared.  Compatibility may be an issue.

jbergstroem commented:
Looks like this changelog entry:
```
2003-01-01  Zack Rusin  <zack@kde.org>

	* main.c (main): added --libs-only-other and --cflags-only-other
	arguments, thanks to which a more obscure dependencies can be
	retrieved, e.g. -pthread
```

jbergstroem commented:
@bnoordhuis I don't mind reverting to the more generic `--cflags`, but since I'm to split on `-I` it felt safer.

bnoordhuis commented:
--cflags-only-I is fine for include_dirs.  I was worried about general compiler flags ending up there.

jbergstroem commented on configure:
```diff
@@ -857,12 +839,13 @@ def configure_intl(o):
     o['variables']['v8_enable_i18n_support'] = 1
     pkgicu = pkg_config('icu-i18n')
-    if not pkgicu:
+    if pkgicu[1] == None:
```
Actually, we probably want to check for libs here. I'll update.

jbergstroem commented:
@bnoordhuis ok, I've tried different configurations of shared libraries on different platforms now (including  icu). Had to convert the output of pkg-config from bytes to string since I had issues with how gyp handled output printing.

bnoordhuis commented on configure:
```diff
@@ -672,41 +685,27 @@ def configure_node(o):
+    default_lib = format_libraries(getattr(options, shared_lib + '_libname'))
+    default_libpath = getattr(options, shared_lib + '_libpath')
+    if default_libpath:
+      default_libpath = "-L" + default_libpath
```
Femto-style nit: can you use single quotes for consistency here and below?

bnoordhuis commented on configure:
```diff
@@ -672,41 +685,27 @@ def configure_node(o):
+    libpath = pkg_libpath if pkg_libpath else default_libpath
+
+    if libs:
+      output['libraries'] += libs.split()
```
If `libs.split()` is always safe here (because it's always in the form of `-lfoo -lbar`) can you add a comment explaining that?

jbergstroem replied:
Ok.

bnoordhuis commented on configure:
```diff
@@ -672,41 +685,27 @@ def configure_node(o):
+    if cflags:
+      output['include_dirs'] += [cflags]
+    if libpath:
+      output['libraries'] += [libpath]
```
Style nit: two newlines.

bnoordhuis commented on configure:
```diff
@@ -857,12 +842,13 @@ def configure_intl(o):
     o['variables']['v8_enable_i18n_support'] = 1
     pkgicu = pkg_config('icu-i18n')
-    if not pkgicu:
+    if pkgicu[0] == None:
```
Idiomatic: `if pkgicu[0] is None:`

bnoordhuis commented on configure:
```diff
@@ -857,12 +842,13 @@ def configure_intl(o):
-    (libs, cflags) = pkgicu
+    (libs, cflags, libpath) = pkgicu
     o['libraries'] += libs.split()
+    o['libraries'] += libpath.split()
```
Can you add a similar comment about `.split()` as requested above?

jbergstroem replied:
Will do, this is wrong though -- libpath needs to be treated as a string similar to pkg-config. I might look at refactoring this to using `configure_library`  instead.

bnoordhuis commented:
LGTM with comments.

mgorny commented on configure:
```diff
@@ -343,17 +343,30 @@ def b(value):
-  return (libs, cflags)
+  pkg_config = 'pkg-config'
+  pkg_config = os.environ.get('PKG_CONFIG', pkg_config)
+  pkg_config = os.environ.get('PKG_CONFIG_PATH', pkg_config)
```
You're clobbering the executable with .pc file search paths :P.

jbergstroem replied:
PKG_CONFIG is in there for backwards compatibility from the floating pkg-config patch and PKG_CONFIG_PATH is suggested by pkg-config upstream.

mgorny replied:
Whaaat? `PKG_CONFIG` is used by some build systems to point to pkg-config executable. `PKG_CONFIG_PATH` contains search path for .pc files.

jbergstroem replied:
I'll re-read the docs. A python library also added it to path which is why I considered it in the first place.

mgorny commented on configure:
```diff
@@ -343,17 +343,30 @@ def b(value):
+
+def format_libraries(list):
+  """Returns string of space separated libraries"""
+  set = list.split(',')
```
`set` is a type name, I'd avoid that.

jbergstroem replied:
good point.

mgorny commented on configure:
```diff
@@ -343,17 +343,30 @@ def b(value):
+def format_libraries(list):
+  """Returns string of space separated libraries"""
+  set = list.split(',')
+  return ' '.join('-l{0}'.format(i) for i in set)
```
You're using `'x' + y` and `'x{0}'.format(y)` inconsistently in the added code. I'd say decide on one :).

jbergstroem replied:
Yeah, was mixed for readability (based on feedback by bnoordhuis). I previously had format everywhere.

mgorny commented on configure:
```diff
@@ -672,40 +685,29 @@ def configure_node(o):
+    default_libpath = getattr(options, shared_lib + '_libpath')
+    if default_libpath:
+      default_libpath = '-L' + default_libpath
+    (pkg_libs, pkg_cflags, pkg_libpath) = pkg_config(lib)
```
This will throw a `ValueError` (or sth like that) when you return `None` instead of the tuple.

jbergstroem replied:
I always return a tuple -- worst case being (`None`, `None`, `None`)

mgorny replied:
My bad, confused `val =` with `retval =`.

mgorny commented on configure:
```diff
@@ -672,40 +685,29 @@ def configure_node(o):
+    if default_libpath:
+      default_libpath = '-L' + default_libpath
+    (pkg_libs, pkg_cflags, pkg_libpath) = pkg_config(lib)
+    cflags = pkg_cflags.split('-I') if pkg_cflags else default_cflags
```
That's not the way you should split stuff :). You're just asking for trouble. Say, I use `/home/Foo-Ibn` prefix.

jbergstroem replied:
How do you suggest I split it?

mgorny replied:
On whitespace, I guess.

jbergstroem replied:
We discussed that previously. What if paths have a space in them? Its a tricky situation. Most userland libraries that parses pkg-config output (there's a few in python, ruby, JavaScript and perl) chooses between `-I` and space. My call (and bnoordhuis suggestion) was that `-I` probably would be less common.

rlidwka replied:
`-I` would also be less expected.

Experienced users usually understand that naming a folders with a space in them is just asking for trouble, so it is commonly avoided. But nobody is explicitly avoiding `-I` in their paths.

mgorny replied:
I'd say ` -I` (space + `-I`) would be a safer compromise. Just prepend a space to get clean split :).

jbergstroem replied:
The problem with splitting with space is that the first item won't strip its `-I`. I'd have to do some extra logic for that to work as intended. I think `-I` is rare enough. If we get a bug filed, I'll revisit.

mgorny replied:
I just told you to prepend a single space, and it will work.

jbergstroem replied:
It won't work for the first result, since it wouldn't split `-Ifoo`

mgorny commented on configure:
```diff
@@ -718,25 +720,11 @@ def configure_v8(o):
   o['variables']['node_shared_openssl'] = b(options.shared_openssl)
-  o['variables']['openssl_no_asm'] = (
-    1 if options.openssl_no_asm else 0)
+  o['variables']['openssl_no_asm'] = 1 if options.openssl_no_asm else 0
```
Any reason not to use `b()` thingie here? Just guessing it goes for boolean, though I can't grep for it right now :P.

jbergstroem replied:
It's not a boolean, its string true or false but the openssl build system requires 1 or 0.

mgorny replied:
Ok, my bad.

mgorny commented on configure:
```diff
@@ -857,12 +845,15 @@ def configure_intl(o):
+    # safe to split, cannot contain spaces
     o['libraries'] += libs.split()
+    # libpath provides linker path which may contain spaces
+    o['libraries'] += [libpath]
```
I'd say logically libpath belongs before libs :).

jbergstroem replied:
Yep, you're right.

mgorny commented on configure:
```diff
@@ -1020,15 +1011,18 @@ if (options.dest_os):
 configure_fullystatic(output)
 
+# remove duplicates from libraries
+output['libraries'] = list(set(output['libraries']))
```
Dangerous. You reorder libraries, you ask for symbol resolution failures.

jbergstroem replied:
Yeah, you're right. Ignore duplicates?

mgorny replied:
Yes. Just pass whatever you get. Don't do magic, duplicates don't hurt.

jbergstroem commented:
Committed in 2b1c01c2cc203dd6131c3fc6238cf62a2e697f63. Thanks for your review, @bnoordhuis  – and thanks for the input @mgorny, @rlidwka.
